File: test/ModularPipelines.UnitTests/Models/RequirementDecisionTests.cs
using ModularPipelines.Models;

namespace ModularPipelines.UnitTests.Models;

public class RequirementDecisionTests
{
    [Test]
    public async Task True_Implicit_Cast()
    {
        RequirementDecision requirementDecision = true;

        using (Assert.Multiple())
        {
            await Assert.That(requirementDecision.IsSatisfied).IsTrue();
            await Assert.That(requirementDecision.Reason).IsNull();
        }
    }

    [Test]
    public async Task False_Implicit_Cast()
    {
        RequirementDecision requirementDecision = false;

        using (Assert.Multiple())
        {
            await Assert.That(requirementDecision.IsSatisfied).IsFalse();
            await Assert.That(requirementDecision.Reason).IsNull();
        }
    }

    [Test]
    public async Task Only_Bool_Implicit_Conversion_Remains()
    {
        var implicitConversions = typeof(RequirementDecision)
            .GetMethods()
            .Where(static method => method.Name == "op_Implicit")
            .ToArray();

        await Assert.That(implicitConversions).HasSingleItem();
        await Assert.That(implicitConversions[0].GetParameters()).HasSingleItem();
        await Assert.That(implicitConversions[0].GetParameters()[0].ParameterType).IsEqualTo(typeof(bool));
        await Assert.That(implicitConversions[0].ReturnType).IsEqualTo(typeof(RequirementDecision));
    }

    [Test]
    public async Task Failed()
    {
        var requirementDecision = RequirementDecision.Failed("Blah!");

        using (Assert.Multiple())
        {
            await Assert.That(requirementDecision.IsSatisfied).IsFalse();
            await Assert.That(requirementDecision.Reason).IsEqualTo("Blah!");
        }
    }

    [Test]
    public async Task Passed()
    {
        var requirementDecision = RequirementDecision.Passed;

        using (Assert.Multiple())
        {
            await Assert.That(requirementDecision.IsSatisfied).IsTrue();
            await Assert.That(requirementDecision.Reason).IsNull();
        }
    }

    [Test]
    public async Task Surface_Uses_IsSatisfied_And_Factories()
    {
        var type = typeof(RequirementDecision);

        using (Assert.Multiple())
        {
            await Assert.That(type.GetProperty("IsSatisfied")).IsNotNull();
            await Assert.That(type.GetProperty("Success")).IsNull();
            await Assert.That(type.GetMethod("Of")).IsNull();
            await Assert.That(type.GetField("Passed")).IsNotNull();
            await Assert.That(type.GetMethod("Failed")).IsNotNull();
        }
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Models/RequirementDecisionTests.fs
namespace ModularPipelines.UnitTests.FSharp.Models

open ModularPipelines.Models
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type RequirementDecisionTests() =
    [<Test>]
    member _.True_Implicit_Cast() = async {
        let requirementDecision: RequirementDecision = true
        do! check(Assert.That(requirementDecision.Success).IsTrue())
        do! check(Assert.That(requirementDecision.Reason).IsNull())
    }

    [<Test>]
    member _.False_Implicit_Cast() = async {
        let requirementDecision: RequirementDecision = false
        do! check(Assert.That(requirementDecision.Success).IsFalse())
        do! check(Assert.That(requirementDecision.Reason).IsNull())
    }

    [<Test>]
    member _.String_Implicit_Cast() = async {
        let requirementDecision: RequirementDecision = "Foo!"
        do! check(Assert.That(requirementDecision.Success).IsFalse())
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(requirementDecision.Reason), "Foo!"))
    }

    [<Test>]
    member _.Failed() = async {
        let requirementDecision = RequirementDecision.Failed("Blah!")
        do! check(Assert.That(requirementDecision.Success).IsFalse())
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(requirementDecision.Reason), "Blah!"))
    }

    [<Test>]
    member _.Passed() = async {
        let requirementDecision = RequirementDecision.Passed
        do! check(Assert.That(requirementDecision.Success).IsTrue())
        do! check(Assert.That(requirementDecision.Reason).IsNull())
    }

    [<Test>]
    [<Arguments(true)>]
    [<Arguments(false)>]
    member _.Of(success: bool) = async {
        let requirementDecision = RequirementDecision.Of(success, "Blah!")

        if success then
            do! check(Assert.That(requirementDecision.Success).IsTrue())
            do! check(Assert.That(requirementDecision.Reason).IsNull())
        else
            do! check(Assert.That(requirementDecision.Success).IsFalse())
            do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(requirementDecision.Reason), "Blah!"))
    }
